Disallow kNN searches with index alias filters

mayya-sharipova opened the pull request:
This PR throws an exception for kNN searches on filtered aliases.
We don't allow kNN searches on filtered aliases as currently filters are
applied only after kNN searches are done, which may lead to returning
less than k results.

In the future, we want to apply filters while doing a kNN search.
Once implemented, we will allow kNN searches on filtered aliases.

Relates to #78473

mayya-sharipova committed:
Disallow kNN searches with index alias filters

This PR throws an exception for kNN searches on filetered aliases.
We don't allow kNN searches on filtered aliases as currently filters are
applied only after kNN searches are done, which may lead to returning
less than k results.

In the future, we want to apply filters while doing a kNN search.
Once implemented, we will allow kNN searches on filtered aliases.

Relates to #78473

mayya-sharipova committed:
Merge remote-tracking branch 'upstream/master' into knn-filter-alias

jtibshirani reviewed (commented):
This approach makes sense to me. Maybe we could double-check if the data management team is okay with it -- it looks like they've done a lot of work on `IndicesOptions` through their work on the index APIs.

mayya-sharipova committed:
Address Julie's feedback

- Move all request building logic to KnnSearchRequestBuilder
- Move a check for allowing filtered aliases to IndicesService

mayya-sharipova committed:
Merge remote-tracking branch 'upstream/master' into knn-filter-alias
